lib/src/components/NxDropdown/AbstractDropdown.tsx

```diff
@@ -0,0 +1,152 @@
+ * @param onCloseKeyDown - A callback provided that should fire when a key press
+ *    is detected that would result in the dropdown getting closed. The Keyboard event is passed to this handler,
+ *    and if the handler calls `preventDefault` on it, the close is cancelled
+ * @param onToggleCollapse - The onToggleCollapse function that was provided from outside of the component
+ * @param onKeyDown - Any onKeyDown function that might have been provided from outside of the component.
+ *    Note that this fires after the close processing occurs
+ * @return - A div with the toggle element and dropdown menu.
+ */
+
+const AbstractDropdown = forwardRef<HTMLDivElement, AbstractDropdownProps>((props: AbstractDropdownProps, ref) => {
+  const {
+    className,
+    isOpen,
+    disabled,
+    toggleElement,
+    children,
+    onToggleCollapse: onToggleCollapseProp,
+    onKeyDown: onKeyDownProp,
+    onCloseClick,
+    onCloseKeyDown: onCloseKeyDownProp,
+    ...attrs
+  } = props;
+
+  const menuRef = useRef<HTMLDivElement>(null);
+  const toggleRef = useRef<HTMLButtonElement>(null);
+
+  const onKeyDown: KeyboardEventHandler<HTMLDivElement> = event => {
+    if (isOpen && !disabled && event.key === 'Escape' && onToggleCollapseProp) {
+      if (onCloseKeyDownProp) {
+        onCloseKeyDownProp(event);
+      }
+
+      if (!event.defaultPrevented) {
+        onToggleCollapseProp();
+
+        // try to prevent the ESC from triggering multiple things, for instance closing a surrounding modal
+        // in addition to the dropdown
+        event.preventDefault();
+      }
+    }
+
+    if (onKeyDownProp) {
+      onKeyDownProp(event);
+    }
+  };
+
+  const handleDocumentClick = (event: MouseEvent) => {
+    if (isOpen && onToggleCollapseProp) {
+      let defaultPrevented = false;
+
+      if (onCloseClick) {
+        /* when consuming code calls preventDefault on the event, we only want that to affect this dropdown.
+        * Since the mouse clicks are captured at the document level, calling preventDefault on the actual
+        * native event could have a wide variety of other effects, so we proxy the event object overriding
+        * preventDefault with our own impl that only affects a locally-scoped variable
+        */
+        const proxiedClickEvent = new Proxy(event, {
+          get(target, prop) {
+            if (prop === 'preventDefault') {
+              return () => { defaultPrevented = true; };
+            }
+            else {
+              return Reflect.get(target, prop);
+            }
+          }
+        });
+
+        onCloseClick(proxiedClickEvent);
+      }
+
+      if (!defaultPrevented) {
+        onToggleCollapseProp();
+      }
+    }
+  };
+
+  function onToggleCollapse() {
+    // When closing, this is handled by handlDocumentClick. So only do it here when opening
+    // (i.e. when not already open)
+    !isOpen && onToggleCollapseProp && onToggleCollapseProp();
```

Comment:
I'm not a huge fan of using `&&` short-circuiting for side-effects, I'd rather use a normal `if` block

Reply:
> I'm not a huge fan of using `&&` short-circuiting for side-effects, I'd rather use a normal `if` block

Updated here: 4112512fc83b79f755931c83d7df51fa75b88751